packages/middleware/rate-limiter/src/sliding-window.ts

```diff
@@ -0,0 +1,125 @@
+import { Rate, RateLiteral, parseRate } from './rate'
+import { Slot, SlotProvider } from './slot'
+import { types } from '@mondrian-framework/model'
+import { functions } from '@mondrian-framework/module'
+
+//inspired by https://blog.cloudflare.com/counting-things-a-lot-of-different-things/
+
+/**
+ * Rate limit at `rateLimit` in the `samplingPeriod`.
+ * Example of maximum 10 request every 2 minutes:
+ * ```typescript
+ * new SlidingWindow({ rateLimit: 10, samplingPeriod: "2 minutes" })
+ * ```
+ */
+export class SlidingWindowProvider {
+  private readonly slidingWindows: Map<string, SlidingWindow> = new Map()
+  private readonly rate: Rate
+  private readonly slotProvider: SlotProvider
+
+  constructor({ rate, slotProvider }: { rate: Rate | RateLiteral; slotProvider: SlotProvider }) {
+    this.rate = typeof rate === 'string' ? parseRate(rate) : rate
+    this.slotProvider = slotProvider
+  }
+
+  get({ key }: { key: string }): SlidingWindow {
+    const slidingWindow = this.slidingWindows.get(key)
+    if (slidingWindow) {
+      return slidingWindow
+    }
+    const newSlidingWindow = new SlidingWindow({ rate: this.rate, slotProvider: this.slotProvider, key })
+    this.slidingWindows.set(key, newSlidingWindow)
+    return newSlidingWindow
+  }
+}
+
+export class SlidingWindow {
+  private readonly samplingPeriodSeconds: number
+  private readonly rateLimit: number
+  private readonly slots: Map<number, Slot> = new Map()
+  private readonly nowSeconds: () => number
+  private readonly slotProvider: SlotProvider
+  private readonly key: string
+  private rateLimitedUntil?: number
+
+  constructor({
+    rate,
+    nowSeconds,
+    slotProvider,
+    key,
+  }: {
+    rate: Rate
+    nowSeconds?: () => number
+    slotProvider: SlotProvider
+    key: string
+  }) {
+    if (rate.period < 1) {
+      throw new Error('Sampling period must be at least 1 second')
+    }
+    if (rate.requests < 0) {
+      throw new Error('Rate limit must be a positive duration')
+    }
+    this.samplingPeriodSeconds = rate.period * (rate.unit === 'hours' ? 3600 : rate.unit === 'minutes' ? 60 : 1)
+    this.rateLimit = rate.requests
+    this.nowSeconds = nowSeconds ?? (() => new Date().getTime() / 1000.0)
+    this.slotProvider = slotProvider
+    this.key = key
+  }
+
+  private getSlots(): [Slot, Slot] {
```

Comment:
`getSlots` is quite a cryptic name, it doesn't really explain anything except the fact that you're returning some slots.
- Why are exactly two slots returned?
- Do these slots have any particular property?
- What are the invariants holding for these slots?

Also this function is not simply returning slots but is also updating the object's internal state by removing references to other slots. I wouldn't expect a simple "getter" function to have such behaviour, its name should reflect that:
`getSlotsXYZAndRemoveOldReferences`.

If you think it looks bad and is too long then the function may be split in two different functions, but anyways I prefer a longer name rather than letting unexpected behaviour slip in my code by accident